client/blocks/signup-form/index.jsx

```diff
@@ -249,6 +251,14 @@ class SignupForm extends Component {
 				: mapKeys( response.messages, ( value, key ) => camelCase( key ) );
 
 			forEach( messages, ( fieldError, field ) => {
+				// If they've yet to tab to the password or username fields, don't show an error message prematurely
+				if ( field === 'password' && ! this.state.focusPassword ) {
+					delete messages.password;
+				}
+				if ( field === 'username' && ! this.state.focusUsername ) {
+					delete messages.username;
+				}
```

Comment:
Looking at the server-side source of the `/signups/validation/user` endpoint that `validateNewUser()` calls, I see it has pretty good support for validating an incomplete signup form. If we took advantage of it here, it would fix the bug in a cleaner way.

The endpoint supports validation of the following fields:
```js
'email', 'first_name', 'last_name', 'username', 'password'
```
If any of them is missing (e.g., we provide only `username`), the endpoint will validate only the supplied fields individually.

Only when the complete set is supplied, the form is validated as a whole using the `wpmu_validate_user_signup` hook.

Therefore, if we used `focusPassword` and `focusUsername` to filter the field set in the _request_ rather than in the _response_ (see the `fieldsForValidation` array a few lines above), the validation endpoint would get more precise input and could provide more precise and relevant output.

Feature-rich React form libraries like Redux Form or Formik support a concept of a "pristine" status of a field. If a field is "pristine", i.e., the user hasn't typed anything into it yet, it can be excluded from validation, as that usually triggers only annoying and irrelevant errors. The `focus*` state fields here provide a basic implementation of that concept. In the long term, we should convert all our forms to use one of the 3rd party libraries.

Reply:
Thanks for pointing that out @jsnajdr! Updated in https://github.com/Automattic/wp-calypso/commit/0c67059c4885b2462ad77c503181dbff312e0caa